app/helpers/smart_proxies_helper.rb

```diff
@@ -74,4 +74,16 @@ def subnet_label(subnet)
       subnet.network
     end
   end
+
+  def link_to_host(subnet, hostname, smart_proxy)
+    foreman_subnet = Subnet.where(:network => subnet.network, :netmask => subnet.netmask)
+    return hostname unless foreman_subnet
+    host = hosts_with_subnet(subnet, smart_proxy, :name => hostname).first
+    host.present? ? link_to(host.to_s, host_path(host)) : hostname
+  end
+
+  def hosts_with_subnet(subnet, smart_proxy, condition = {})
+    subnets =  { :subnets => { :dhcp_id => smart_proxy.id, :network => subnet.network } }
+    Host.joins(:interfaces => :subnet).where(subnets.merge! condition)
```

Comment:
imho you should keep it in an instance, as otherwise rails will still general this query for each host (even if cached)?